File: golang/linked-list/easy/83.删除排序链表中的重复元素.go
/*
 * @lc app=leetcode.cn id=83 lang=golang
 * @lcpr version=30202
 *
 * [83] 删除排序链表中的重复元素
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
	dummy := &ListNode{Val: -200}
	dummy.Next = head
	slow, fast := dummy, dummy
	for fast != nil {
		if slow.Val != fast.Val {
			slow.Next = fast
			slow = fast
		}

		fast = fast.Next
	}

	slow.Next = nil
	return dummy.Next
}

// @lc code=end

/*
// @lcpr case=start
// [1,1,2]\n
// @lcpr case=end

// @lcpr case=start
// [1,1,2,3,3]\n
// @lcpr case=end

*/

File: cpp/linked-list/easy/83.删除排序链表中的重复元素.cpp
/*
 * @lc app=leetcode.cn id=83 lang=cpp
 *
 * [83] 删除排序链表中的重复元素
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode* deleteDuplicates(ListNode* head)
    {
        if (head == nullptr)
            return nullptr;

        auto slow = head;
        auto fast = head;
        while (fast != nullptr) {
            if (slow->val != fast->val) {
                slow->next = fast;
                slow       = fast;
            }
            fast = fast->next;
        }
        slow->next = nullptr;

        return head;
    }
};
// @lc code=end
